### PR TITLE
Make purge_history operate on tokens

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/storage/events.py [hunk 1]
```diff
@@ -2109,10 +2119,20 @@ def _purge_history_txn(
         #
         # So, let's stick it at the end so that we don't block event
         # persistence.
-        logger.info("[purge] updating room_depth")
+        txn.execute("""
+            SELECT COALESCE(MIN(depth), 0)
+            FROM event_backward_extremities AS eb
+            INNER JOIN event_edges AS eg ON eg.prev_event_id = eb.event_id
+            INNER JOIN events AS e ON e.event_id = eg.event_id
```
Comment (richvdh) [inline synapse/storage/events.py]: is this looking for the prev_event before the extremity? which we have likely just purged?

File: synapse/storage/events.py [hunk 2]
```diff
@@ -1927,9 +1932,8 @@ def _purge_history_txn(
         txn.execute(
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: How so?

File: synapse/storage/events.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/storage/events.py]: This is looking at the events that point to the extremity, as we don't actually have the extremity events.

Comment (erikjohnston) [reply synapse/storage/events.py]: (And I think the check is the right way round?)

File: synapse/storage/events.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/storage/events.py]:
```
# We calculate the new entries for the backward extremeties by finding
# all events that point to events that are to be purged
```

we now look for events we are about to purge, which point to events we are *not* going to purge. it's completely different.

File: synapse/storage/events.py (hunk 1 above)
Comment (richvdh) [reply synapse/storage/events.py]: oh right. ehm. I'm not convinced this is going to calculate the right thing in the presence of disconnected bits of dag.

File: synapse/storage/events.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/events.py]: OK, its different in terms that we're not returning the event that points to purged events but the purged events that we get pointed at, yes. (And this isn't anything to do with the change in the sql, and was wrong before)

File: synapse/storage/events.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/storage/events.py]: Each disconnected part of the DAG will have their backwards extremities in the `event_backward_extremities` table, so the event with the minimum depth must be one of them, no?

**Pull request merged**